sapi/fpm/fpm/zlog.c

```diff
@@ -414,7 +414,8 @@ static inline ssize_t zlog_stream_unbuffered_write(
 static inline ssize_t zlog_stream_buf_copy_cstr(
 		struct zlog_stream *stream, const char *str, size_t str_len) /* {{{ */
 {
-	if (stream->buf.size - stream->len <= str_len && !zlog_stream_buf_alloc_ex(stream, str_len)) {
+	if (stream->buf.size - stream->len <= str_len &&
+			!zlog_stream_buf_alloc_ex(stream, str_len + stream->len)) {
```

Comment:
What about the zlog_stream_buf_alloc_ex use in the next function? Doesn't it have the same problem?

The MIN(zlog_limit) in the allocation function also looks quite dangerous to me and requires careful management of invariants.

I'm wondering if it would make sense to move zlog over to use smart_string instead of a homegrown implementation.

Reply:
> What about the zlog_stream_buf_alloc_ex use in the next function? Doesn't it have the same problem?

There are just two and the first one passes 0 which means allocate `buf_init_size` and the second one is 1 which will most likely double the buffer from the quick look so no problem there I think.

> The MIN(zlog_limit) in the allocation function also looks quite dangerous to me and requires careful management of invariants.

There's never a point to allocate more than zlog_limit and it's bounded to at least 512 so I don't see issue with that.

> I'm wondering if it would make sense to move zlog over to use smart_string instead of a homegrown implementation.

I remember I needed something a bit more custom as there are some special requirements with wrapping, padding and so on. It seemed better to use special implementation for that and think I didn't like something on smart string but don't remember what was that. Also I didn't want external deps (core php ones) as those are limited just to some parts of fpm. Not really big issue though but in any case, it's not really worth any time on changing that as this works well and there are some extra protections - that's why this is not really a bug that would cause any problem because it's a path that cannot happen.